File: CFNDeploy/python/cfndeploy/app_stack.py
"""
Application infrastructure stack.
"""
import aws_cdk as cdk
from aws_cdk import aws_ec2 as ec2, aws_autoscaling as autoscaling, aws_codedeploy as codedeploy, aws_iam as iam
from constructs import Construct

class AppStack(cdk.Stack):
    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)
        
        # Create VPC
        vpc = ec2.Vpc(self, "AppVPC", max_azs=2, nat_gateways=1)
        
        # Create security group
        security_group = ec2.SecurityGroup(
            self, "AppSecurityGroup",
            vpc=vpc,
            description="Security group for application instances",
            allow_all_outbound=True
        )
        
        security_group.add_ingress_rule(ec2.Peer.any_ipv4(), ec2.Port.tcp(80), "Allow HTTP")
        security_group.add_ingress_rule(ec2.Peer.any_ipv4(), ec2.Port.tcp(22), "Allow SSH")
        
        # Create IAM role
        instance_role = iam.Role(
            self, "InstanceRole",
            assumed_by=iam.ServicePrincipal("ec2.amazonaws.com"),
            managed_policies=[
                iam.ManagedPolicy.from_aws_managed_policy_name("CloudWatchAgentServerPolicy"),
                iam.ManagedPolicy.from_aws_managed_policy_name("AmazonSSMManagedInstanceCore")
            ]
        )
        
        # Create launch template
        launch_template = ec2.LaunchTemplate(
            self, "AppLaunchTemplate",
            instance_type=ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.MICRO),
            machine_image=ec2.MachineImage.latest_amazon_linux2(),
            security_group=security_group,
            role=instance_role,
            user_data=ec2.UserData.for_linux()
        )
        
        launch_template.user_data.add_commands(
            "yum update -y",
            "yum install -y httpd ruby wget",
            "systemctl start httpd",
            "systemctl enable httpd",
            "echo '<h1>Hello from CDK Pipeline!</h1>' > /var/www/html/index.html",
            "cd /home/ec2-user",
            "wget https://aws-codedeploy-us-east-1.s3.us-east-1.amazonaws.com/latest/install",
            "chmod +x ./install",
            "./install auto"
        )
        
        # Create Auto Scaling Group
        asg = autoscaling.AutoScalingGroup(
            self, "AppASG",
            vpc=vpc,
            launch_template=launch_template,
            min_capacity=1,
            max_capacity=3,
            desired_capacity=2,
            vpc_subnets=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS)
        )
        
        # Create CodeDeploy application
        application = codedeploy.ServerApplication(
            self, "AppApplication",
            application_name=f"my-app-{self.stack_name}"
        )
        
        # Create deployment group
        deployment_group = codedeploy.ServerDeploymentGroup(
            self, "AppDeploymentGroup",
            application=application,
            deployment_group_name="production",
            auto_scaling_groups=[asg],
            install_agent=True,
            ec2_instance_tags=codedeploy.InstanceTagSet({"Environment": ["production"]})
        )
        
        # Outputs
        cdk.CfnOutput(self, "VPCId", value=vpc.vpc_id, description="VPC ID")
        cdk.CfnOutput(self, "ApplicationName", value=application.application_name, description="CodeDeploy Application Name")
        cdk.CfnOutput(self, "DeploymentGroupName", value=deployment_group.deployment_group_name, description="CodeDeploy Deployment Group Name")

File: JenkinsBuild/javascript/lib/app-stack.js
const cdk = require('aws-cdk-lib');
const ec2 = require('aws-cdk-lib/aws-ec2');
const autoscaling = require('aws-cdk-lib/aws-autoscaling');
const codedeploy = require('aws-cdk-lib/aws-codedeploy');
const iam = require('aws-cdk-lib/aws-iam');

class AppStack extends cdk.Stack {
  constructor(scope, id, props) {
    super(scope, id, props);

    // Create VPC
    const vpc = new ec2.Vpc(this, 'AppVPC', {
      maxAzs: 2,
      natGateways: 1,
    });

    // Create security group
    const securityGroup = new ec2.SecurityGroup(this, 'AppSecurityGroup', {
      vpc,
      description: 'Security group for application instances',
      allowAllOutbound: true,
    });

    securityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(80),
      'Allow HTTP traffic'
    );

    securityGroup.addIngressRule(
      ec2.Peer.anyIpv4(),
      ec2.Port.tcp(22),
      'Allow SSH access'
    );

    // Create IAM role for EC2 instances
    const instanceRole = new iam.Role(this, 'InstanceRole', {
      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('CloudWatchAgentServerPolicy'),
        iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore'),
      ],
    });

    // Create instance profile
    const instanceProfile = new iam.CfnInstanceProfile(this, 'InstanceProfile', {
      roles: [instanceRole.roleName],
    });

    // Create launch template
    const launchTemplate = new ec2.LaunchTemplate(this, 'AppLaunchTemplate', {
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.T3, ec2.InstanceSize.MICRO),
      machineImage: ec2.MachineImage.latestAmazonLinux2(),
      securityGroup,
      role: instanceRole,
      userData: ec2.UserData.forLinux(),
    });

    launchTemplate.userData.addCommands(
      'yum update -y',
      'yum install -y httpd',
      'systemctl start httpd',
      'systemctl enable httpd',
      'echo "<h1>Hello from CDK Pipeline!</h1>" > /var/www/html/index.html',
      'yum install -y ruby wget',
      'cd /home/ec2-user',
      'wget https://aws-codedeploy-us-east-1.s3.us-east-1.amazonaws.com/latest/install',
      'chmod +x ./install',
      './install auto'
    );

    // Create Auto Scaling Group
    const asg = new autoscaling.AutoScalingGroup(this, 'AppASG', {
      vpc,
      launchTemplate,
      minCapacity: 1,
      maxCapacity: 3,
      desiredCapacity: 2,
      vpcSubnets: {
        subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
      },
    });

    // Create CodeDeploy application
    const application = new codedeploy.ServerApplication(this, 'AppApplication', {
      applicationName: `my-app-${this.stackName}`,
    });

    // Create CodeDeploy deployment group
    const deploymentGroup = new codedeploy.ServerDeploymentGroup(this, 'AppDeploymentGroup', {
      application,
      deploymentGroupName: 'production',
      autoScalingGroups: [asg],
      installAgent: true,
      ec2InstanceTags: new codedeploy.InstanceTagSet({
        'Environment': ['production'],
      }),
    });

    // Outputs
    new cdk.CfnOutput(this, 'VPCId', {
      value: vpc.vpcId,
      description: 'VPC ID',
    });

    new cdk.CfnOutput(this, 'ApplicationName', {
      value: application.applicationName,
      description: 'CodeDeploy Application Name',
    });

    new cdk.CfnOutput(this, 'DeploymentGroupName', {
      value: deploymentGroup.deploymentGroupName,
      description: 'CodeDeploy Deployment Group Name',
    });
  }
}

module.exports = { AppStack };